### ISSUE TITLE
Bug in multi-parental name referencing

### ISSUE THREAD

**Issue by tfwnicholson** [GPflow/GPflow#349]
I encountered a bug in the naming logic. In the following, quite reasonable, code the third line will throw a `ValueError('This Param appears to be doubly referenced by a parent')`:
```
constant_mean = Constant(1)
additive_mean = constant_mean + constant_mean
constant_mean_name = constant_mean.name
```
The `additive_mean` clearly contains two references to the `constant_mean`, but this seems to me that it should be valid. 

I see a few possible solutions:
 + allowing multi parental references (why was this disallowed in the first place?)
 + detecting this case in the constructor and raising there
 + detecting this case in the constructor and replacing the additive mean with a: `Constant(2) * constant_mean`

Which method would you prefer? I'll prepare a PR with it.

**Comment by jameshensman** [comment]
I think the desired behaviour is that we shouldn't allow multiple references to parameters, so the code you posted should raise an exception. 

To fix we should examine parameters (and parameterizeds) being assigned to the parameterzed class, and reject them if we already have a reference to them. In fact, we should reject them if there is a reference to them anywhere on the tree.

 This should result in the above code raising an exception in the constructor of `Add()`, which is called when the __add__ function is called using the `+` operator.

**Comment by alexggmatthews** [comment]
Fixed in #364 #370
